### PR TITLE
chunk coin_store request into smaller sql queries

### PR DESCRIPTION
**Pull request by arvidn**
to not exceed the limit of 999 on old versions of sqlite.

I also took the opportunity to factor out the `chunks()` function from weight proof into its own file and added a simple unit test.

There's indentation change in the `CoinStore` functions, you may want to review those ignoring white space.

### CODE REVIEW

File: chia/full_node/coin_store.py [hunk 1]
```diff
@@ -311,24 +314,25 @@ async def get_coin_states_by_puzzle_hashes(
...
-            f"AND (confirmed_index>=? OR spent_index>=?)"
-            f"{'' if include_spent_coins else 'AND spent_index=0'}",
-            puzzle_hashes_db + (min_height, min_height),
-        ) as cursor:
+        for puzzles in chunks(puzzle_hashes, MAX_SQLITE_PARAMETERS):
+            puzzle_hashes_db: Tuple[Any, ...]
+            if self.db_wrapper.db_version == 2:
+                puzzle_hashes_db = tuple(puzzles)
+            else:
+                puzzle_hashes_db = tuple([ph.hex() for ph in puzzles])
```
Comment (altendky) [inline chia/full_node/coin_store.py:322]: This is a change unrelated to this PR, but maybe catch it while touching it.

```suggestion
                puzzle_hashes_db = tuple(ph.hex() for ph in puzzles)
```

File: chia/full_node/coin_store.py [hunk 2]
```diff
@@ -341,23 +345,24 @@ async def get_coin_records_by_parent_ids(
...
-            for row in await cursor.fetchall():
-                coin = self.row_to_coin(row)
-                coins.add(CoinRecord(coin, row[0], row[1], row[2], row[6]))
-            return list(coins)
+        for ids in chunks(parent_ids, MAX_SQLITE_PARAMETERS):
+            parent_ids_db: Tuple[Any, ...]
+            if self.db_wrapper.db_version == 2:
+                parent_ids_db = tuple(ids)
+            else:
+                parent_ids_db = tuple([pid.hex() for pid in ids])
```
Comment (altendky) [inline chia/full_node/coin_store.py:353]: As above, this is unrelated to the change.  Ignore if you want it in a separate PR.

```suggestion
                parent_ids_db = tuple(pid.hex() for pid in ids)
```

File: chia/full_node/coin_store.py [hunk 3]
```diff
@@ -369,22 +374,22 @@ async def get_coin_states_by_ids(
...
-
-            for row in await cursor.fetchall():
-                coins.add(self.row_to_coin_state(row))
-            return list(coins)
+        for ids in chunks(coin_ids, MAX_SQLITE_PARAMETERS):
+            coin_ids_db: Tuple[Any, ...]
+            if self.db_wrapper.db_version == 2:
+                coin_ids_db = tuple(ids)
+            else:
+                coin_ids_db = tuple([pid.hex() for pid in ids])
```
Comment (altendky) [inline chia/full_node/coin_store.py:382]: likewise

```suggestion
                coin_ids_db = tuple(pid.hex() for pid in ids)
```

File: chia/full_node/coin_store.py [hunk 4]
```diff
@@ -311,24 +314,25 @@ async def get_coin_states_by_puzzle_hashes(
...
+            async with self.coin_record_db.execute(
+                f"SELECT confirmed_index, spent_index, coinbase, puzzle_hash, "
+                f"coin_parent, amount, timestamp FROM coin_record INDEXED BY coin_puzzle_hash "
+                f'WHERE puzzle_hash in ({"?," * (len(puzzles) - 1)}?) '
+                f"AND (confirmed_index>=? OR spent_index>=?)"
+                f"{'' if include_spent_coins else 'AND spent_index=0'}",
+                puzzle_hashes_db + (min_height, min_height),
+            ) as cursor:
+
+                async for row in cursor:
```
Comment (altendky) [inline chia/full_node/coin_store.py:332]: I think I like the middle ground of multiple `.fetchmany()` calls in general.  But, maybe in this form we just want to have the batching be based on the desired fetch batch size rather than the limit of the max query params?  Like maybe we want a batch size of 100 or something and go back to `.fetchall()`?

File: chia/full_node/coin_store.py [hunk 5]
```diff
@@ -7,11 +7,14 @@
 from chia.util.db_wrapper import DBWrapper
 from chia.util.ints import uint32, uint64
 from chia.util.lru_cache import LRUCache
+from chia.util.chunks import chunks
```
Comment (altendky) [inline chia/full_node/coin_store.py:10]:
```suggestion
from chia.util.chunks import chunks
from chia.util.db_wrapper import DBWrapper
from chia.util.ints import uint32, uint64
from chia.util.lru_cache import LRUCache
```

File: chia/full_node/weight_proof.py [hunk 6]
```diff
@@ -20,6 +20,7 @@
     calculate_sp_iters,
     is_overflow_block,
 )
+from chia.util.chunks import chunks
```
Comment (altendky) [inline chia/full_node/weight_proof.py:23]: Belongs down a bit, GitHub won't let me make a suggestion for this one.

File: chia/wallet/wallet_node.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 import traceback
 from asyncio import CancelledError
 from pathlib import Path
-from typing import Callable, Dict, List, Optional, Set, Tuple, Any
+from typing import Callable, Dict, List, Optional, Set, Tuple, Any, Iterator
```
Comment (altendky) [inline chia/wallet/wallet_node.py:9]:
```suggestion
from typing import Callable, Dict, Iterator, List, Optional, Set, Tuple, Any
```

File: chia/wallet/wallet_node.py [hunk 8]
```diff
@@ -21,7 +21,7 @@
     KeychainProxy,
     KeyringIsEmpty,
 )
-from chia.full_node.weight_proof import chunks
+from chia.util.chunks import chunks
```
Comment (altendky) [inline chia/wallet/wallet_node.py:24]: Belongs lower down.

File: tests/util/test_chunks.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def test_chunks() -> None:
+
+    assert list(chunks([], 0)) == []
+    assert list(chunks(["a"], 0)) == [["a"]]
+    assert list(chunks(["a", "b"], 0)) == [["a"], ["b"]]
+
+    assert list(chunks(["a", "b", "c", "d"], 2)) == [["a", "b"], ["c", "d"]]
+    assert list(chunks(["a", "b", "c", "d"], 200)) == [["a", "b", "c", "d"]]
```
Comment (altendky) [inline tests/util/test_chunks.py:11]:
```suggestion
    assert list(chunks(["a", "b", "c", "d"], 4)) == [["a", "b", "c", "d"]]
    assert list(chunks(["a", "b", "c", "d"], 200)) == [["a", "b", "c", "d"]]

    assert list(chunks(["a", "b", "c", "d"], 3)) == [["a", "b", "c"], ["d"]]
```

**Review by altendky — CHANGES_REQUESTED** [review]
Only really needed change is the extra test cases at the end.  The rest are just comments that you may or may not want to include.  The import order will otherwise get addressed when we reduce the isort ignore list size.  The unneeded `[]` are pre-existing.

File: chia/full_node/weight_proof.py (hunk 6 above)
Comment (arvidn) [reply chia/full_node/weight_proof.py:23]: does `isort` not fix this?
If not, what rule are we following?

File: chia/full_node/coin_store.py (hunk 1 above)
Comment (arvidn) [reply chia/full_node/coin_store.py:322]: I was under the impression list-comprehension was specific to lists, and that you had to include the `[` and `]`. is the expression `f(x) for x in y` producing an iterator?
but then, the expression `[x]` is a list of `x`, even if `x` is an iterator, right?

File: chia/full_node/coin_store.py (hunk 4 above)
Comment (arvidn) [reply chia/full_node/coin_store.py:332]: I would like to keep it simple until we measure it to be too slow. It probably is a bit slower with an `async for` since it will (probably) need to check the future every turn through the loop, but I wouldn't think it makes a big difference.

**Comment by arvidn** [comment]
I'm hesitant to manually apply `isort` rules. I think it would make more sense to remove some files from the exemption. But maybe not mixed in here.

**Comment by arvidn** [comment]
I added the test cases

File: chia/full_node/coin_store.py (hunk 1 above)
Comment (altendky) [reply chia/full_node/coin_store.py:322]: > I was under the impression list-comprehension was specific to lists, and that you had to include the `[` and `]`.

There are list comprehensions but also dict and set comprehensions.  Plus generator comprehensions (previously referred to as generator expressions).

> is the expression `f(x) for x in y` producing an iterator?

A small caveat.  `(f(x) for x in y)` is a generator comprehension and creates a generator which does indeed implement the iterator protocol.  If you pass it to a function as the only parameter then you can drop the parentheses for the generator comprehension and just use the existing function call parentheses like in `a_function(f(x) for x in y)`, or `tuple(f(x) for x in y)` as in the code here.

> but then, the expression `[x]` is a list of `x`, even if `x` is an iterator, right?

Yes.  To add to that, if you want to create a list and iterate it to get the elements of the list you would instead `list(the_iterable)`.

File: chia/full_node/weight_proof.py (hunk 6 above)
Comment (altendky) [reply chia/full_node/weight_proof.py:23]: `isort` is configured to ignore presently-not-compliant files.  mypy has a similar setup.  We just have giant lists of exceptions basically so new files will be clean and we can go through existing files one at a time, clean them, then remove them from the exceptions list.

**Comment by altendky** [comment]
> I'm hesitant to manually apply `isort` rules. I think it would make more sense to remove some files from the exemption. But maybe not mixed in here.

I wouldn't do the whole thing.  I would encourage a habit of putting new imports where they belong so there's less delta when we swing through with dedicated `isort` PRs.  Remember that concerns about order being changed in an automated fashion is why we have exceptions to begin with.  Keeping new additions tidy along the way will make the cleanup easier and safer to get merged.

But yes, I'm not pushing for a change here.

**Review by altendky — APPROVED** [review]
